core/src/main/java/org/infinispan/commands/write/RemoveExpiredCommand.java

```diff
@@ -85,15 +91,27 @@ public Object perform(InvocationContext ctx) throws Throwable {
          } else if (e.getLifespan() > 0 && e.getLifespan() == lifespan) {
             // If the entries lifespan is not positive that means it can't expire so don't even try to remove it
             // Lastly if there is metadata we have to verify it equals our lifespan and the value match.
-            // TODO: add a threshold to verify this wasn't just created with the same value/lifespan just before expiring
             if (valueMatcher.matches(prevValue, value, null)) {
-               e.setExpired(true);
-               return performRemove(e, prevValue, ctx);
+               // Only remove the entry if it is still expired.
+               // Due to difference in system clocks between nodes - we also accept the expiration within 100 ms
+               // of now. This along with the fact that entries are not normally access immediately when they
+               // expire should give a good enough buffer range to not create a false positive.
+               if (ExpiryHelper.isExpiredMortal(lifespan, e.getCreated(), timeService.wallClockTime() + 100)) {
```

Comment:
Can you make this into a constant. Hardcoding is fine, but I like the visibility at the top of the class